File: golang/3622.check-divisibility-by-digit-sum-and-product.go
func checkDivisibility(n int) bool {
  var d int
  
  p, s := 1, 0

  for x := n; x > 0; x /= 10 {
    d = x % 10
    p *= d
    s += d
  }

  return n % (s + p) == 0
}

File: rust/3622.check-divisibility-by-digit-sum-and-product.rs
use std::iter::successors;

impl Solution {
    pub fn check_divisibility(n: i32) -> bool {
        let (sum, prod) = successors(Some(n), |&x| if x > 0 { Some(x / 10) } else { None })
            .take_while(|&x| x > 0)
            .map(|x| x % 10)
            .fold((0, 1), |(s, p), d| (s + d, p * d));

        n % (sum + prod) == 0
    }
}
